lxd/storage_volumes.go

```diff
@@ -62,11 +62,81 @@ var storagePoolVolumesTypeCmd = APIEndpoint{
-	Get:    APIEndpointAction{Handler: storagePoolVolumeGet, AccessHandler: allowPermission(entity.TypeStorageVolume, auth.EntitlementCanView, "poolName", "type", "volumeName")},
-	Patch:  APIEndpointAction{Handler: storagePoolVolumePatch, AccessHandler: allowPermission(entity.TypeStorageVolume, auth.EntitlementCanEdit, "poolName", "type", "volumeName")},
-	Post:   APIEndpointAction{Handler: storagePoolVolumePost, AccessHandler: allowPermission(entity.TypeStorageVolume, auth.EntitlementCanEdit, "poolName", "type", "volumeName")},
-	Put:    APIEndpointAction{Handler: storagePoolVolumePut, AccessHandler: allowPermission(entity.TypeStorageVolume, auth.EntitlementCanEdit, "poolName", "type", "volumeName")},
+	Delete: APIEndpointAction{Handler: storagePoolVolumeDelete, AccessHandler: storagePoolVolumeAccessHandler(auth.EntitlementCanEdit)},
+	Get:    APIEndpointAction{Handler: storagePoolVolumeGet, AccessHandler: storagePoolVolumeAccessHandler(auth.EntitlementCanView)},
+	Patch:  APIEndpointAction{Handler: storagePoolVolumePatch, AccessHandler: storagePoolVolumeAccessHandler(auth.EntitlementCanEdit)},
+	Post:   APIEndpointAction{Handler: storagePoolVolumePost, AccessHandler: storagePoolVolumeAccessHandler(auth.EntitlementCanEdit)},
+	Put:    APIEndpointAction{Handler: storagePoolVolumePut, AccessHandler: storagePoolVolumeAccessHandler(auth.EntitlementCanEdit)},
+}
+
+// storagePoolVolumeAccessHandler resolves the project and location of the storage volume before performing the permission check.
+func storagePoolVolumeAccessHandler(entitlement auth.Entitlement) func(d *Daemon, r *http.Request) response.Response {
+	return func(d *Daemon, r *http.Request) response.Response {
+		s := d.State()
+		poolName, err := url.PathUnescape(mux.Vars(r)["poolName"])
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		volumeTypeName, err := url.PathUnescape(mux.Vars(r)["type"])
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		volumeType, err := storagePools.VolumeTypeNameToDBType(volumeTypeName)
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		volumeName, err := url.PathUnescape(mux.Vars(r)["volumeName"])
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		location := request.QueryParam(r, "target")
+
+		projectName := request.ProjectParam(r)
+		err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
+			dbProject, err := dbCluster.GetProject(ctx, tx.Tx(), projectName)
+			if err != nil {
+				return err
+			}
+
+			apiProject, err := dbProject.ToAPI(ctx, tx.Tx())
+			if err != nil {
+				return err
+			}
+
+			projectName = project.StorageVolumeProjectFromRecord(apiProject, volumeType)
+
+			// For storage volume we also need to populate its location.
+			if location == "" {
+				storagePoolID, err := tx.GetStoragePoolID(ctx, poolName)
```

Comment:
I'm concerned about the duplication (and thus increase) in the amount of DB queries these (and earlier changes in this PR) are doing, especially when taken in the context that we are already concerned about the increase in DB queries the openfga driver is going to be performing.

We do need to take the proliferation of DB queries seriously I'm afraid.
They are not cheap when made from non-leader members of a cluster.

If we take, for example, `storagePoolVolumeDelete` handler, as well as the queries performed in this new access handler:

```go
	// Get the storage pool the storage volume is supposed to be attached to.
	pool, err := storagePools.LoadByName(s, poolName)
	if err != nil {
		return response.SmartError(err)
	}

	// Get the storage volume.
	var dbVolume *db.StorageVolume
	err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
		dbVolume, err = tx.GetStoragePoolVolume(ctx, pool.ID(), volumeProjectName, volumeType, volumeName, true)
		return err
	})
	if err != nil {
		return response.SmartError(err)
	}
```

Then we have (at least):
1. dbCluster.GetProject: Get project name and ID
2. dbProject.ToAPI: Get project's config
3.  tx.GetStoragePoolID: Get storage pool ID from name
4. tx.GetStoragePoolVolume: Get storage volume & config (2 queries)
5. s.Authorizer.CheckPermission: Unknown number of queries  (multiple queries)
6. project.StorageVolumeProject: Duplicate of dbCluster.GetProject and dbProject.ToAPI (2 queries) *duplicate*
7. storagePools.LoadByName: Get pool's ID, config and nodes (3 queries) *duplicate*
8.  tx.GetStoragePoolVolume: Get storage volume & config (2 queries) *duplicate*

So over 12 queries just to delete a volume (before we consider the extra for openfga) several of which are duplicated by this and the earlier changes in this PR.

Is there a way we can do the access checking in the handler function itself so we can avoid the duplication?